### PR TITLE
[FIR]: add support for vararg.

### PR DESCRIPTION
**Pull request by juan-chen**
* add varargElementType to FirValueParameter
* add FirVarargArgumentExpression
* add translation of vararg info from Psi -> Fir -> Ir

### CODE REVIEW

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt [hunk 1]
```diff
@@ -659,6 +663,12 @@ class FirRenderer(builder: StringBuilder) : FirVisitorVoid() {
         lambdaArgumentExpression.expression.accept(this)
     }
 
+    override fun visitVarargArgumentExpression(varargArgumentExpression: FirVarargArgumentExpression) {
+        print("vararg(")
```
[the remark this answers is not in the document]
Comment (mglukhikh) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt]: NB: I fixed all test data myself (see https://github.com/JetBrains/kotlin/commit/d13071a3b577b12880d14378364969b5537e9ae9)

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrDeclarationStorage.kt [hunk 2]
```diff
@@ -686,7 +686,8 @@ class Fir2IrDeclarationStorage(
                 IrValueParameterImpl(
                     startOffset, endOffset, origin, symbol,
                     valueParameter.name, index, type,
-                    null, valueParameter.isCrossinline, valueParameter.isNoinline
+                    valueParameter.varargElementType?.toIrType(session, this),
```
Comment (semoro) [inline compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrDeclarationStorage.kt]: You could extract varargElementType from `type` by using `org.jetbrains.kotlin.fir.resolve.calls.ArgumentsKt#varargElementType`

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt [hunk 3]
```diff
@@ -189,7 +192,40 @@ class FirCallCompletionResultsWriterTransformer(
             }
             else -> {
                 resultType = typeRef.substituteTypeRef(subCandidate)
-                result.transformArguments(this, subCandidate.createArgumentsMapping()).transformExplicitReceiver(integerApproximator, null)
+                val vararg = subCandidate.argumentMapping?.values?.firstOrNull { it.isVararg }
+                result.transformArguments(this, subCandidate.createArgumentsMapping()).apply {
+                    if (vararg != null && this is FirFunctionCallImpl) {
+                        // Create a FirVarargArgumentExpression for the vararg arguments
+                        val resolvedArrayType = vararg.returnTypeRef.substitute(subCandidate)
+                        if (resolvedArrayType.typeArguments.size == 1) {
```
Comment (semoro) [inline compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt]: This array might be primitive, so it will not have a generic parameter
`IntArray` for example, use `org.jetbrains.kotlin.fir.resolve.calls.ArgumentsKt#varargElementType`

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt [hunk 4]
```diff
@@ -102,7 +103,7 @@ private class VarargLowering(val context: JvmBackendContext) : FileLoweringPass,
         context.createJvmIrBuilder(currentScope!!.scope.scopeOwnerSymbol, startOffset, endOffset)
 
     private val IrFunctionSymbol.isArrayOf: Boolean
-        get() = this == context.ir.symbols.arrayOf || owner.isPrimitiveArrayOf
+        get() = this == context.ir.symbols.arrayOf || owner.isArrayOf
```
Comment (semoro) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: Looks like it is already checking is this is `kotlin.Array`, so it is no need to check it by name

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt [hunk 5]
```diff
@@ -29,11 +31,15 @@ object ArrayIterator : IntrinsicMethod() {
         signature: JvmMethodSignature,
         context: JvmBackendContext
     ): IrIntrinsicFunction {
-        val owner = context.typeMapper.mapClass(expression.symbol.owner.parentAsClass)
+        val isPrimitive = ((expression.type as? IrSimpleType)?.arguments?.size ?: 0) == 0
+        val owner =
+            if (isPrimitive) context.typeMapper.mapClass(expression.symbol.owner.parentAsClass)
+            else context.typeMapper.mapClass(context.irBuiltIns.arrayClass.owner)
+
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt]: Could you please clarify why these changes in the backend are necessary? Right now this intrinsic works correctly in the old frontend mode. Is there some difference in the IR here? In that case I believe it's best to fix it in fir2ir.

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/ConversionUtils.kt [hunk 6]
```diff
@@ -89,6 +89,7 @@ fun ConeKotlinType.toIrType(
 
 private fun getPrimitiveArrayType(classId: ClassId?, irBuiltIns: IrBuiltIns): IrClassifierSymbol? {
     val irType = when (classId) {
+        ClassId(FqName("kotlin"), FqName("Array"), false) -> return irBuiltIns.arrayClass
```
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/ConversionUtils.kt]: done

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt]: Thanks! I've added d13071a into the PR.

File: compiler/fir/tree/gen/org/jetbrains/kotlin/fir/expressions/FirVarargArgumentExpression.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import org.jetbrains.kotlin.fir.FirSourceElement
+import org.jetbrains.kotlin.fir.types.FirTypeRef
+import org.jetbrains.kotlin.fir.visitors.*
+
+/*
+ * This file was generated automatically
+ * DO NOT MODIFY IT MANUALLY
+ */
+
+abstract class FirVarargArgumentExpression : FirPureAbstractElement(), FirExpression {
```
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/tree/gen/org/jetbrains/kotlin/fir/expressions/FirVarargArgumentExpression.kt]: done

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt [hunk 8]
```diff
@@ -199,14 +235,19 @@ class FirCallCompletionResultsWriterTransformer(
...
+        coneTypeUnsafe<ConeKotlinType>()
+            .let { candidate.substitutor.substituteOrSelf(it) }
+            .let { finalSubstitutor.substituteOrSelf(it) }
+
     private fun Candidate.createArgumentsMapping(): ExpectedArgumentType? {
         return argumentMapping?.map { (argument, valueParameter) ->
-                val expectedType = valueParameter.returnTypeRef.coneTypeUnsafe<ConeKotlinType>()
-                    .let { substitutor.substituteOrSelf(it) }
-                    .let { finalSubstitutor.substituteOrSelf(it) }
+            val expectedType = valueParameter.returnTypeRef.coneTypeUnsafe<ConeKotlinType>()
```
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt]: done

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt [hunk 9]
```diff
@@ -199,14 +235,19 @@ class FirCallCompletionResultsWriterTransformer(
...
+                .let { substitutor.substituteOrSelf(it) }
+                .let { finalSubstitutor.substituteOrSelf(it) }
 
-                argument.expandArgument() to expectedType
-            }
+            argument.expandArgument() to expectedType
```
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt]: done

File: compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrDeclarationStorage.kt (hunk 2 above)
Comment (juan-chen) [reply compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrDeclarationStorage.kt]: done

File: compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt (hunk 3 above)
Comment (juan-chen) [reply compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/FirCallCompletionResultsWriterTransformer.kt]: done

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt (hunk 5 above)
Comment (juan-chen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt]: I'll look for a fix in fir2ir and amend the PR.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt (hunk 4 above)
Comment (juan-chen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: Fir doesn't use the arrayOf symbol and therefore needs check by name.

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt]: I removed varargElementType from FirValueParameter based on Simon's suggestion, and thus need to regenerate the render result. Would you please let me know what's the best way to regenerate?

**Comment by juan-chen** [comment]
I've removed varargElementType from FirValueParameter and used ArrayUtils::arrayElementType to compute the vararg element type. More tests are passing, but there is one regression, varargInJava, which I'm looking at now.

File: compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mglukhikh) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt]: I usually run all tests from `compiler:fir:resolve` and look how much results are changed. If there are not too much of them, I edit data by hand using "click to see difference". If not, I delete all txt files in directory by file system command and then re-run all tests to regenerate test data.

[the remark this answers is not in the document]
Comment (juan-chen) [reply compiler/fir/tree/src/org/jetbrains/kotlin/fir/FirRenderer.kt]: Thanks for the instructions. I added updated testData.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt (hunk 5 above)
Comment (juan-chen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/ArrayIterator.kt]: Restored this file.

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt [hunk 10]
```diff
@@ -111,16 +112,18 @@ private class VarargLowering(val context: JvmBackendContext) : FileLoweringPass,
...
+        private val IrFunction.isArrayOf: Boolean
             get() {
                 val parent = when (val directParent = parent) {
                     is IrClass -> directParent.getPackageFragment() ?: return false
                     is IrPackageFragment -> directParent
                     else -> return false
                 }
                 return parent.fqName == KotlinBuiltIns.BUILT_INS_PACKAGE_FQ_NAME &&
-                        name.asString() in PRIMITIVE_ARRAY_OF_NAMES &&
+                        (name.asString() in PRIMITIVE_ARRAY_OF_NAMES || name.toString() == ARRAY_OF_NAME) &&
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: `toString` -> `asString`

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt (hunk 4 above)
Comment (udalov) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: But then isn't the first condition (`this == context.ir.symbols.arrayOf`) unnecessary anymore?

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt (hunk 10 above)
Comment (juan-chen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: done

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt (hunk 4 above)
Comment (juan-chen) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/VarargLowering.kt]: removed the first condition

**Comment by mglukhikh** [comment]
Merged manually 👍
